File: player4.cs
using System;
using System.Threading;

enum PlayerStatus
{
    Normal,
    Stopped,
    SlowMotion,
    Invisible,
    SpeedBoost,
    Blocked
}

class Player
{
    public string Name { get; private set; }
    public PlayerStatus Status { get; private set; }
    public int[] Position { get; private set; } = new int[2]; // [x, y]

    public Player(string name)
    {
        Name = name;
        Status = PlayerStatus.Normal;
        Position[0] = 0; // X position
        Position[1] = 0; // Y position
    }

    public void MoveLeft()
    {
        if (Status == PlayerStatus.Blocked)
        {
            Console.WriteLine($"{Name} is blocked and cannot move left.");
        }
        else if (Status == PlayerStatus.Stopped)
        {
            Console.WriteLine($"{Name} is stopped and cannot move left.");
        }
        else
        {
            Position[0] -= 1;
            Console.WriteLine($"{Name} moves left. New position: ({Position[0]}, {Position[1]})");
        }
    }

    public void MoveRight()
    {
        if (Status == PlayerStatus.Blocked)
        {
            Console.WriteLine($"{Name} is blocked and cannot move right.");
        }
        else if (Status == PlayerStatus.Stopped)
        {
            Console.WriteLine($"{Name} is stopped and cannot move right.");
        }
        else
        {
            Position[0] += 1;
            Console.WriteLine($"{Name} moves right. New position: ({Position[0]}, {Position[1]})");
        }
    }

    public void MoveForward()
    {
        if (Status == PlayerStatus.Blocked)
        {
            Console.WriteLine($"{Name} is blocked and cannot move forward.");
        }
        else if (Status == PlayerStatus.Stopped)
        {
            Console.WriteLine($"{Name} is stopped and cannot move forward.");
        }
        else
        {
            Position[1] += 1;
            Console.WriteLine($"{Name} moves forward. New position: ({Position[0]}, {Position[1]})");
        }
    }

    public void MoveBackward()
    {
        if (Status == PlayerStatus.Blocked)
        {
            Console.WriteLine($"{Name} is blocked and cannot move backward.");
        }
        else if (Status == PlayerStatus.Stopped)
        {
            Console.WriteLine($"{Name} is stopped and cannot move backward.");
        }
        else
        {
            Position[1] -= 1;
            Console.WriteLine($"{Name} moves backward. New position: ({Position[0]}, {Position[1]})");
        }
    }

    public void Jump()
    {
        if (Status == PlayerStatus.Blocked)
        {
            Console.WriteLine($"{Name} is blocked and cannot jump.");
        }
        else if (Status == PlayerStatus.Stopped)
        {
            Console.WriteLine($"{Name} is stopped and cannot jump.");
        }
        else
        {
            Console.WriteLine($"{Name} jumps!");
            // Simulate jump by moving up and forward
            Position[1] += 2; // Simulate upward jump
            Position[0] += 2; // Simulate forward movement
            Console.WriteLine($"{Name} is at position: ({Position[0]}, {Position[1]}) - Jumping");
            Thread.Sleep(500); // Short delay to simulate jump duration
            Position[1] -= 2; // Simulate landing
            Console.WriteLine($"{Name} lands. New position: ({Position[0]}, {Position[1]})");
        }
    }
}

class Program
{
    static void Main(string[] args)
    {
        Player player = new Player("Martin");

        player.MoveLeft();   // Move left
        player.Jump();       // Player jumps
        player.MoveRight();  // Move right
        player.Jump();       // Player jumps again
    }
}

File: player4.py
import time
from enum import Enum

class PlayerStatus(Enum):
    NORMAL = "normal"
    STOPPED = "stopped"
    SLOW_MOTION = "in slow motion"
    INVISIBLE = "invisible"
    SPEED_BOOST = "speed++"
    BLOCKED = "blocked by an obstacle"

class Player:
    def __init__(self, name):
        self.name = name
        self.status = PlayerStatus.NORMAL
        self.position = [0, 0]  # (x, y) coordinates, y being vertical

    def move_left(self):
        if self.status == PlayerStatus.BLOCKED:
            print(f"{self.name} is blocked and cannot move left.")
        if self.status == PlayerStatus.STOPPED:
            print(f"{self.name} is stopped and cannot move left.")
        else:
            self.position[0] -= 1
            print(f"{self.name} moves left. New position: {self.position}")

    def move_right(self):
        if self.status == PlayerStatus.BLOCKED:
            print(f"{self.name} is blocked and cannot move right.")
        if self.status == PlayerStatus.STOPPED:
            print(f"{self.name} is stopped and cannot move right.")
        else:
            self.position[0] += 1
            print(f"{self.name} moves right. New position: {self.position}")

    def move_forward(self):
        if self.status == PlayerStatus.BLOCKED:
            print(f"{self.name} is blocked and cannot move forward.")
        if self.status == PlayerStatus.STOPPED:
            print(f"{self.name} is stopped and cannot move forward.")
        else:
            self.position[1] += 1
            print(f"{self.name} moves forward. New position: {self.position}")

    def move_backward(self):
        if self.status == PlayerStatus.BLOCKED:
            print(f"{self.name} is blocked and cannot move backward.")
        if self.status == PlayerStatus.STOPPED:
            print(f"{self.name} is stopped and cannot move backward.")
        else:
            self.position[1] -= 1
            print(f"{self.name} moves backward. New position: {self.position}")

    def jump(self):  # Saute en hauteur de 2, en avant de 2 et redescends en hauteur de 2  
        if self.status == PlayerStatus.BLOCKED:
            print(f"{self.name} is blocked and cannot jump.")
        if self.status == PlayerStatus.STOPPED:
            print(f"{self.name} is stopped and cannot jump.")
        else:
            print(f"{self.name} jumps!")
            # Simulate jump by moving up and down on the Y axis
            self.position[1] += 2  # Simulate upward jump
            self.position[0] += 2  # Simulate forward jump
            print(f"{self.name} is at position: {self.position} - Jumping")
            time.sleep(0.5)  # Short delay to simulate the jump
            self.position[1] -= 2  # Simulate landing
            print(f"{self.name} lands. New position: {self.position}")

# Example usage
def main():
    player = Player("Martin")

    player.move_left()   # Move left
    player.jump()        # Player jumps
    player.move_right()  # Move right
    player.jump()        # Player jumps again

if __name__ == "__main__":
    main()
